File: 76.minimum-window-substring.go
/*
 * @lc app=leetcode id=76 lang=golang
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
func minWindow(s string, t string) string {
	if s == "" || t == "" {
		return ""
	}

	var sFreq, tFreq [256]int
	result, left, right, minw, finalLeft, finalRight, count := "", 0, -1, len(s)+1, -1, -1, 0

	for i := 0; i < len(t); i++ {
		tFreq[t[i]-'a']++
	}

	for left < len(s) {
		if right+1 < len(s) && count < len(t) {
			char := s[right+1]
			sFreq[char-'a']++
			if sFreq[char-'a'] <= tFreq[char-'a'] {
				count++
			}
			right++
		} else {
			if right-left+1 < minw && count == len(t) {
				minw = right - left + 1
				finalLeft = left
				finalRight = right
			}

			char := s[left]
			if sFreq[char-'a'] == tFreq[char-'a'] {
				count--
			}
			sFreq[char-'a']--
			left++
		}
	}

	if minw != len(s)+1 {
		result = string(s[finalLeft : finalRight+1])
	}

	return result
}

// @lc code=end

File: 76.minimum-window-substring.cpp
/*
 * @lc app=leetcode id=76 lang=cpp
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
class Solution {
 public:
  string minWindow(string s, string t) {
    map<char, int> s_feq;
    map<char, int> t_feq;
    for (char c : t) ++t_feq[c];

    string ans = "";
    int cnt = 0;
    int min_w = s.size() + 1;
    int l = 0, r = 0;

    while (l < s.size()) {
      // [l, r)
      if (r < s.size() && cnt < t.size()) {
        char c = s[r];
        ++s_feq[c];

        if (s_feq[c] <= t_feq[c])
          ++cnt;

        ++r;
      } else {
        if (cnt == t.size() && r - l < min_w) {
          min_w = r - l;
          ans = s.substr(l, r - l);
        }

        char c = s[l];
        if (s_feq[c] == t_feq[c])
          --cnt;

        --s_feq[c];
        ++l;
      }
    }

    return ans;
  }
};
// @lc code=end

